Inspect Data Values Schema in OpenAPI v3.0.x format

pivotaljohn opened the pull request.

cari-lynn committed:
test the openapi map

Signed-off-by: Varsha Munishwar <vmunishwar@vmware.com>

vmunishwar committed:
Add --data-values-schema-inspect flag

Signed-off-by: Garrett Cheadle <gcheadle@vmware.com>

gcheadle-vmware committed:
Create OpenAPI schema output for scalars and maps

Signed-off-by: Varsha Munishwar <vmunishwar@vmware.com>

cari-lynn committed:
Accept file formats and schema type for --output

Signed-off-by: John Ryan <jtigger@infosysengr.com>

jtigger committed:
(refactor) extract inspectSchema()

cari-lynn committed:
Error when -o=openapi-v3 and not inspecting schema

- updated help text
- when no schema output type is specified, the value of the flag is
  "none"

Signed-off-by: John Ryan <jtigger@infosysengr.com>

jtigger committed:
(refactor) move OpenAPI export to "schema" pkg

jtigger committed:
Document new exported methods/functions

jtigger committed:
Add license header to newly added test

jtigger committed:
Add name for generated schema in OpenAPI schema component

- `component.schemas` must be a map from the name of the schema to its
  definition.

cppforlife commented on pkg/cmd/template/cmd.go:
```diff
@@ -128,6 +133,17 @@ func (o *Options) RunWithFiles(in Input, ui ui.UI) Output {
+	if o.DataValuesFlags.InspectSchema {
+		return o.inspectSchema(schema)
+	}
+	schemaType, err := o.RegularFilesSourceOpts.OutputType.Schema()
```
nit: newline

cppforlife commented on pkg/cmd/template/cmd.go:
```diff
@@ -128,6 +133,17 @@ func (o *Options) RunWithFiles(in Input, ui ui.UI) Output {
+	if err != nil {
+		return Output{Err: err}
+	}
+	if schemaType == regularFilesOutputTypeOpenAPI {
```
should this type be public?

cppforlife commented on pkg/cmd/template/cmd.go:
```diff
@@ -128,6 +133,17 @@ func (o *Options) RunWithFiles(in Input, ui ui.UI) Output {
+		return Output{Err: err}
+	}
+	if schemaType == regularFilesOutputTypeOpenAPI {
+		return Output{Err: errors.New("Output type currently only supported for Data Values Schema (i.e. include --data-values-schema-inspect)")}
```
errors.New -> fmt.Errorf

cppforlife commented on pkg/cmd/template/cmd.go:
```diff
@@ -151,6 +163,31 @@ func (o *Options) RunWithFiles(in Input, ui ui.UI) Output {
+			},
+		}
+	}
+	return Output{Err: fmt.Errorf("Data Values Schema export only supported in OpenAPI v3 format; specify format with --output=%s flag",
```
nit: "Data Values Schema" compare capitalization with other error msgs

cari-lynn replied:
Data values and data values schema is lowercase everywhere in the codebase. Updating to lowercase

cppforlife commented on pkg/cmd/template/regular_input.go:
```diff
@@ -113,3 +126,63 @@ If you want to include those results, use the --output-files or --dangerous-empt
+
+var regularFilesOutputFormatTypes = []string{regularFilesOutputTypeYAML, regularFilesOutputTypeJSON, regularFilesOutputTypePos}
+var regularFilesOutputSchemaTypes = []string{regularFilesOutputTypeOpenAPI}
+var regularFilesOutputTypes = append(regularFilesOutputFormatTypes, regularFilesOutputSchemaTypes...)
```
nit: multiline var

var (
   ...
)

cppforlife commented on pkg/cmd/template/regular_input.go:
```diff
@@ -6,37 +6,47 @@ package template
 
-	cmd.Flags().StringVarP(&s.OutputType, "output", "o", regularFilesOutputTypeYAML, "Output type (yaml, json, pos)")
+	s.OutputType.Types =
+		cmd.Flags().StringSliceP("output", "o", []string{regularFilesOutputTypeYAML},
```
use StringSliceVarP

cppforlife commented on pkg/cmd/template/schema_inspect_test.go:
```diff
@@ -0,0 +1,119 @@
+	expected := `openapi: 3.0.0
+info:
+  version: 0.1.0
+  title: Openapi schema generated from ytt Data Values Schema
```
TODO figure out better title

cppforlife commented on pkg/cmd/template/schema_inspect_test.go:
```diff
@@ -0,0 +1,119 @@
+paths: {}
+components:
+  schemas:
+    dataValues:
```
TODO figure better name

cari-lynn replied:
We gave this some thought to this name and arrived at dataValues is an appropriate name for now. It does still leave open the possibility to make this configurable later on.

cppforlife commented on pkg/schema/openapi.go:
```diff
@@ -0,0 +1,84 @@
+			{Key: "version", Value: "0.1.0"},
+			{Key: "title", Value: "Openapi schema generated from ytt Data Values Schema"},
+		}},
+		},
```
nit: indent is weird looking.

pivotaljohn replied:
We tried a number of other formats — all of which _reduce_ readability of the actual content: being able to see the key names _as they appear_ in the resulting OpenAPI document is particularly handy to visually parse this code.

- gofmt uses this format.
- go compiler catches misplaced curly braces.

cppforlife commented on pkg/schema/openapi.go:
```diff
@@ -0,0 +1,84 @@
+	case *DocumentType:
+		return calculateProperties(typedValue.GetValueType())
+	case *MapType:
+		typeString := convertTypeToString(typedValue.String())
```
can we avoid string=> string mapping here and go to type=>string

cppforlife commented on pkg/schema/openapi.go:
```diff
@@ -0,0 +1,84 @@
+			{Key: "type", Value: typeString},
+			{Key: "default", Value: defaultVal},
+		}}
+		if typedValue.String() == "float" {
```
can we use a type here instead of string check?

cppforlife commented on pkg/schema/openapi.go:
```diff
@@ -0,0 +1,84 @@
+			newMap.Items = append(newMap.Items, &yamlmeta.MapItem{Key: "format", Value: "float"})
+		}
+		return &newMap
+
```
nit: remove newline

cppforlife commented on pkg/schema/openapi.go:
```diff
@@ -0,0 +1,84 @@
+
+// NewOpenAPIDocument generates an AST of an OpenAPI v3.0.x document, populating the `schemas:` section with the
+//   type information contained in `docType`.
+func NewOpenAPIDocument(docType *DocumentType) *yamlmeta.Document {
```
instead of having few functions floating in a package, lets turn this into a 

```
type OpenAPIDocument struct {}
func NewOpenAPIDocument(docType *DocumentType) OpenAPIDocument
func (OpenAPIDocument) AsDoc() *yamlmeta.Document
```

pivotaljohn replied:
done.

cppforlife commented on pkg/schema/openapi.go:
```diff
@@ -0,0 +1,84 @@
+	case "boolean":
+		return "boolean"
+	case "integer":
+		return "integer"
```
does this work for int64, uint ... https://github.com/vmware-tanzu/carvel-ytt/blob/develop/pkg/template/core/go_value.go#L67

pivotaljohn replied:
We updated this method to match on the actual type (and not the string name) 👍🏻 .

In terms of handling variants of types...

Scalar types are normalized here:
https://github.com/vmware-tanzu/carvel-ytt/blob/8d6339a9fab27d06ea1a56f32ae24248aedbadbd/pkg/schema/schema.go#L214-L221

which (currently) means there are four scalar "types" to detect.

That said, the precision of the underlying value is preserved. We're using the `ValueType` field to carry the rough type information. E.g. for an integer, the `ValueType` = `int` but the underlying value may be `uint64`.

While we want to handle more types, this handles the input we get from Starlark.

cppforlife commented on pkg/cmd/template/regular_input.go:
```diff
@@ -6,37 +6,47 @@ package template
-)
-
+// RegularFilesSourceOpts holds configuration for when regular files are the input/output
+//   of the execution (via the --files and --output... flags).
```
nit: remove leading 2 spaces.

pivotaljohn replied:
Looked at other projects: no leading spaces on subsequent lines.

Done.

cppforlife commented on pkg/cmd/template/regular_input.go:
```diff
@@ -113,3 +126,63 @@ If you want to include those results, use the --output-files or --dangerous-empt
+var regularFilesOutputTypes = append(regularFilesOutputFormatTypes, regularFilesOutputSchemaTypes...)
+
+// Format returns which of the file format types is in effect
+func (o *OutputType) Format() (string, error) {
```
TODO(dk): check out after simplifying output type cobra flag handling.

cari-lynn committed:
Export output type constants

- use StringSliceVarP instead of StringSliceP

Signed-off-by: John Ryan <jtigger@infosysengr.com>

cari-lynn committed:
Introduce OpenAPIDocument type

- to scope all functions related to converting from ytt schema
- converting scalar types based on the actual value's type (and not a
  string)

Signed-off-by: John Ryan <jtigger@infosysengr.com>
Signed-off-by: Cari Lynn <cdean@pivotal.io>

cari-lynn committed:
Update title message

cppforlife commented on pkg/cmd/template/cmd.go:
```diff
@@ -128,6 +132,18 @@ func (o *Options) RunWithFiles(in Input, ui ui.UI) Output {
 	}
 
+	if o.DataValuesFlags.InspectSchema {
+		return o.inspectSchema(schema)
```
nit: inspectSchema -> inspectDataValuesSchema

cari-lynn replied:
Since the flag is on the DataValuesFlags, it might be redundant to change the name.

cppforlife commented on pkg/schema/openapi.go:
```diff
@@ -0,0 +1,90 @@
+		return "string"
+	case float64:
+		return "number"
+	case int:
+		return "integer"
```
not clear what happens with uint etc.

cari-lynn replied:
We normalize the inputs for ints here

https://github.com/vmware-tanzu/carvel-ytt/blob/8d6339a9fab27d06ea1a56f32ae24248aedbadbd/pkg/schema/schema.go#L218-L22

We don't consider uint there now, while we should add this, I think it belongs in a separate story since it is related to schema, and not OpenAPI.

cppforlife commented on pkg/schema/openapi.go:
```diff
@@ -0,0 +1,90 @@
+	default:
+		panic(fmt.Sprintf("Unrecognized type %T", schemaVal))
+	}
+	return nil
```
nit: this could be dropped afaik

cari-lynn committed:
Remove unnecessary return

 Signed-off-by: John Ryan <jtigger@infosysengr.com>
